### ISSUE TITLE
feat: Ensure the libwasmedge will not expose unnecessary symbols

### ISSUE THREAD

**Issue by hydai** [WasmEdge/WasmEdge#3743]
### Summary

Please check the previous issue. #3655 
WasmEdge made a regression by exposing lots of unnecessary symbols in 0.14.0 and fixed this regression in 0.14.1.
However, we would like to avoid this problem happening again.

### Details

Thanks to the ideas provided by the author of #3655 (Faidon Liambotis). We currently have the following solutions:

1. Use `dpkg-gensymbols` to check if we expose any unexpected symbols.
2. Another way is to use an allowed list to show the explicit symbol list instead of the wildcard. Ref: https://github.com/WasmEdge/WasmEdge/issues/3655#issuecomment-2291097273

Moreover, we welcome any good ideas for fixing this. If you have any thoughts, please feel free to comment below. Thanks!

**Comment by devesh0099** [comment]
I want to work on this. I think we can go with the `dpkg-gensymbols`  solution of detecting exposed symbols in build this way we can catch all the symbols that can be made visible from a file other than libwasmedge.lds and also this way we can have a whitelist of symbols that need to be exposed in a separate file. In future, contributors only need to update this single file separately from other codebase to expose their symbols/functions.

If there is some other better method to deal with this I can try that too.

**Comment by hydai** [comment]
Sounds great! We can maintain the allowed list for the exposed functions.

**Comment by hydai** [comment]
Hi @devesh0099 
Any updates?

**Comment by devesh0099** [comment]
Hey,
I tried to implement the `dpkg-gensymbols `in the GitHub actions for automatically detecting but failed to do so and haven't been able to solve this yet...

**Comment by hydai** [comment]
Do you need any assistance? Could you describe the issue you encountered during the setup of GitHub Actions with `dpkg-gensymbols`?

**Comment by devesh0099** [comment]
Yes please, I am confused about how to properly add a workflow. How do I add a post-build action properly in the workflow directory for all the different builds?

About my post-build action:
I have created a whitelist for the symbol file for the current version if other collaborators want to expose more functions they will have to update the file with the function name. Currently, I decided to keep this file inside `/lib/api`

- Action will generate a symbol file using the ''nm" command 
- Will extract only the name of the function from the generated file
- Check using the "diff" command for any difference
- If found -> fail the build
- If not -> delete the file and continue other actions.

Please share some feedback if you think this is a good solution.

**Comment by hydai** [comment]
I don't think we should use `nm` and parse the function names manually. Using `dpkg-gensymbols` is a better way since it already handles this use case.

**Comment by devesh0099** [comment]
`dpkg-gensymbols` also appends some extra metadata beside function name like

`dpkg-gensymbols` generate:
`WasmEdge_ASTModuleListExports@Base 0.2-1`

`nm` generate:
`0000000000053460 T WasmEdge_ASTModuleListExports`

and for matching we only required 
`WasmEdge_ASTModuleListExports`

**Comment by devesh0099** [comment]
Using `dpkg-gensymbols` results in the same problem in that it requires a complete `debian` file structure under `debian` folder like:

`debian\changelog` and `debian\control`

and I generated those using `dh_make` command but that resulted in too many extra files and folders for just generating symbol files.

I also have a question:
If in future someone choose to remove an exposed function and resulted in whitelist.symbols having extra symbols would CI automatically changes the file or tell the collaborator to change it?

**Comment by hydai** [comment]
I see. Then using `nm` is acceptable for me.

> If in future someone choose to remove an exposed function and resulted in whitelist.symbols having extra symbols would CI automatically changes the file or tell the collaborator to change it?

The person who wants to change the exposed functions should update the allowed list at the same time.

**Comment by devesh0099** [comment]
Do I need to implement this checking for all the builds in GitHub actions?
I am able to successfully implement this check for 
`resuable-build-on-debian.yml`
`resuable-build-on-fedora.yml`
`resuable-build-on-ubuntu.yml`

I will try for Windows and MacOS also

besides these, there are static versions of the distros, RISCV, manylinux, and Android 
these yml files don't contain a `Test Wasmedge` step so what to do in the case of these `.yml` files?

**Comment by hydai** [comment]
We can leave those distros alone.
The focus is on linux(debian, fedora), windows, and macos because we are shipping the packages on the package management systems such as apt, dnf, winget, and homebrew.

**Comment by hydai** [comment]
Hi @devesh0099 
Are you still working on this issue?

**Comment by devesh0099** [comment]
Hey, 
Sorry but I am not solving it now, I was stuck at resolving the mac part of it, sorry for not reporting.

**Comment by Ytemiloluwa** [comment]
Hello @hydai I will like to take up this issue as an applicant for the LFX 25 term 1. I have experience setting up GitHub workflows and I can leverage that to fix this issue. kindly assign this issue to me. Thank you.

**Comment by hydai** [comment]
Assigned. It's yours now.

**Comment by Ytemiloluwa** [comment]
> Assigned. It's yours now.

Okay, thank you.

**Comment by Ytemiloluwa** [comment]
Hello @hydai I have been trying to build wasmedge from source on Mac ( Sequoia ) so far only 26 tests have passed out of 28. the two tests that failed are : 

3:wasmedgeLLVMCoreTests
15:wasmedgeAPIAOTCoreTests 

is this a common issue? is there something I can do to solve it or is it because I am on a Mac (Sequoia). your feedback is appreciated so I can speed up my PR for this issue.

**Comment by hydai** [comment]
> 3:wasmedgeLLVMCoreTests 15:wasmedgeAPIAOTCoreTests

It should not fail, our CI will grant this. You can try to dump more logs when you run the tests.

**Comment by IITI-tushar** [comment]
@hydai 
i would like to work on this!!!
Please assign

**Comment by hydai** [comment]
Fixed by #4471
